ofrak_core/test_ofrak/components/test_patch_symbol_resolution.py

```diff
@@ -0,0 +1,148 @@
+import pytest
+import tempfile
+
+from dataclasses import dataclass
+from immutabledict import immutabledict
+from typing import Dict
+
+from ofrak_patch_maker.toolchain.llvm_12 import LLVM_12_0_1_Toolchain
+from ofrak.core.architecture import ProgramAttributes
+from ofrak_type.architecture import (
+    InstructionSet,
+)
+from ofrak_patch_maker.model import AssembledObject
+from ofrak_patch_maker.patch_maker import PatchMaker
+from ofrak_patch_maker.toolchain.model import (
+    CompilerOptimizationLevel,
+    BinFileType,
+    Segment,
+    ToolchainConfig,
+)
+from ofrak_type.bit_width import BitWidth
+from ofrak_type.endianness import Endianness
+from ofrak_type.memory_permissions import MemoryPermissions
+from ofrak_type.symbol_type import LinkableSymbolType
+
+PARAMS = [
+    {"rel_symbol": {}, "expected_value": 0},
```

Comment:
Thank you for writing these unit tests! It's mostly good, however it needs to be improved in two ways: Stricter coverage over more cases, and readability.

This test should check calculated set of unresolved symbols is exactly equal to what we expect, not just the number of them. It should also have more tests cases, at a minimum:
* Multiple object files using the same symbol, one object file defining the symbol
* Multiple object files defining the same symbol, one object file using the symbol (should not error at this stage, that's the linker's job, but the symbol should be correctly identified as resolved)
* One object file using the symbol, one defining it, one or more not defining or using it

The reason why I say these other cases need to be added at a minimum is that the problems I saw in previous iterations of the `_resolve_symbols_within_BOM`-logic (before it was in a function) arose especially from cases with several object files, with multiple using the symbol. Those errors would not be caught by this unit test, so the test does not test the logic sufficiently.

I feel it's hard to understand what each case is testing at the moment, so some straightforward suggestions to make it more readable are:
* Rename the symbols in the test to be more meaningful, e.g. "defined_in_bom" instead of "foo"
* Add a string label to each test case, explaining it like a comment would. The nice thing about a string label is you can also have pytest use those labels to name each test case in the `parametrize` decorator by adding an argument to parametrize, e.g. `ids=lambda tc: tc.label`(see attached screenshot for example of how that looks).
* Rename the fields of the dictionary to be more understandable at-a-glance, e.g. "rel_symbol" -> "symbols_referenced"

<img width="535" alt="Screen Shot 2023-03-13 at 6 59 53 PM" src="https://user-images.githubusercontent.com/17018663/224851100-32f89cd2-34e2-4d31-bcab-f3330c31e0c6.png">